BREAKING CHANGE: AjvValidator does not support ajv instances anymore

adriasala124 opened the pull request:
AjvValidator instances can be configured fully through the options

Commit:
feat: remove support for passing ajv instances to AjvValidator

BREAKING CHANGE: AjvValidator does not support ajv instances anymore
AjvValidator instances can be configured fully through the options

Commit:
chore(example): add examples

Commit:
chore(docs): document ajv validation

joseandrespg commented on examples/api-express/src/index.ts:
```diff
@@ -7,13 +7,24 @@ import { createApp } from '@davinci/core';
+				header: { coerceTypes: true },
+				query: { coerceTypes: true }
+			},
+			plugins: [[addErrors as any]]
```
Why do we need to cast to any? I imagine plugins would have a common interface or base type

adriasala124 replied:
For some reason TS is not getting the types correctly. I asked @pierissimo about this and he will take a look soon

pierissimo replied:
probably a mismatch between the Ajv version installed as dependency of the `api-express` example, and the Ajv version installed as devDependency in `http-server`

adriasala124 replied:
The version difference is just a patch and by the changelog I don't see anything relevant.
I think it could be a problem of the ajv-error plugin, since I tried with the ajv-formats plugin and it works in this case

adriasala124 replied:
You're right @pierissimo. Bumping ajv in the examples from 8.11.0 to 8.11.2 fixed the problem

Commit:
fix: no need to cast addErrors to any

Commit:
fix(examples): bump ajv deps in examples to 8.11.2 to fix typescript problem with ajv plugin types

pierissimo commented on packages/http-server/src/AjvValidator.ts:
```diff
@@ -20,19 +20,21 @@ const defaultAjvOptions: Options = {
-	ajvOptions?: Options;
-	plugins?: Array<[AjvPlugin, AjvPluginOptions?]>;
-	instances?: Ajv | Partial<AjvInstancesMap>;
+	ajvOptions?: Options | Partial<AjvOptionsMap>;
```
I would rename this to `options`, for consistency.
Or both to `ajvOptions` and `ajvPlugins`, either is fine.

pierissimo commented on packages/http-server/src/AjvValidator.ts:
```diff
@@ -155,41 +158,36 @@ export class AjvValidator<Request = unknown> {
+					const [plugin, opts] = p;
+					plugin(this.ajvInstances[source], opts);
+				});
+			} else {
```
```suggestion
			} else if (Array.isArray(plugins)) {
```

pierissimo commented on packages/http-server/src/AjvValidator.ts:
```diff
@@ -155,41 +158,36 @@ export class AjvValidator<Request = unknown> {
+					plugin(this.ajvInstances[source], opts);
+				});
+			} else {
+				(plugins as AjvPlugins).forEach(p => {
```
```suggestion
				plugins.forEach(p => {
```

pierissimo commented on packages/http-server/src/AjvValidator.ts:
```diff
@@ -155,41 +158,36 @@ export class AjvValidator<Request = unknown> {
+		sources.forEach(source => {
+			if (this.isPluginsMap(plugins)) {
+				// eslint-disable-next-line no-unused-expressions
+				(plugins[source] as AjvPlugins)?.forEach(p => {
```
```suggestion
				plugins[source]?.forEach(p => {
```

pierissimo committed:
chore: small amendments

- Added validation link in the sidebar of docusaurus
- Documentation changes on Validation
- Renamed `plugins` to `ajvPlugins`
- Removed `as` operator
